File: bot/src/docs/export.ts
import { execFile } from "node:child_process";
import { mkdtemp, readFile, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { promisify } from "node:util";

const execFileAsync = promisify(execFile);

/** Supported Pandoc output formats (DESIGN §R3). */
export type ExportFormat = "docx" | "pdf";

export class ExportError extends Error {
  constructor(
    public readonly code: "PANDOC_UNAVAILABLE" | "EMPTY" | "PANDOC_FAILED" | "INVALID_FORMAT",
    message: string,
  ) {
    super(message);
    this.name = "ExportError";
  }
}

const FORMAT_META: Record<
  ExportFormat,
  { ext: string; mime: string; pandocArgs: string[] }
> = {
  docx: {
    ext: ".docx",
    mime: "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    pandocArgs: [],
  },
  pdf: {
    ext: ".pdf",
    mime: "application/pdf",
    pandocArgs: [],
  },
};

let pandocAvailable: boolean | null = null;

/** Test hook — reset cached pandoc probe between unit tests. */
export function resetPandocProbe(): void {
  pandocAvailable = null;
}

/** True when `pandoc` is on PATH and responds to `--version`. */
export async function isPandocAvailable(): Promise<boolean> {
  if (pandocAvailable != null) return pandocAvailable;
  try {
    await execFileAsync("pandoc", ["--version"], { timeout: 5000 });
    pandocAvailable = true;
  } catch {
    pandocAvailable = false;
  }
  return pandocAvailable;
}

/** Drop the deterministic citation footer before export or doctrine save. */
export function stripSourcesFooter(text: string): string {
  const marker = "\n\n📎 Sources:";
  const idx = text.indexOf(marker);
  return idx >= 0 ? text.slice(0, idx) : text;
}

/** Normalize markdown body for Pandoc (strip chat footers, trim). */
export function prepareMarkdownForExport(markdown: string): string {
  return stripSourcesFooter(markdown).trim();
}

/** Suggested download filename from a doctrine source path. */
export function exportFilename(source: string, format: ExportFormat): string {
  const base =
    source
      .replace(/\\/g, "/")
      .replace(/\.(md|markdown)$/i, "")
      .split("/")
      .pop() || "document";
  return `${base}${FORMAT_META[format].ext}`;
}

export function parseExportFormat(raw: unknown): ExportFormat | null {
  if (raw === "docx" || raw === "pdf") return raw;
  return null;
}

/**
 * Convert markdown to docx/pdf via Pandoc. Writes temp files under os.tmpdir();
 * never persists exports on disk beyond the call.
 */
export async function exportMarkdown(markdown: string, format: ExportFormat): Promise<Buffer> {
  if (!(await isPandocAvailable())) {
    throw new ExportError("PANDOC_UNAVAILABLE", "pandoc is not installed or not on PATH");
  }

  const body = prepareMarkdownForExport(markdown);
  if (!body) throw new ExportError("EMPTY", "document is empty");

  const meta = FORMAT_META[format];
  const dir = await mkdtemp(join(tmpdir(), "moneypenny-export-"));
  const inPath = join(dir, "input.md");
  const outPath = join(dir, `output${meta.ext}`);

  try {
    await writeFile(inPath, body, "utf8");
    await execFileAsync("pandoc", [inPath, "-o", outPath, ...meta.pandocArgs], {
      timeout: 60_000,
      maxBuffer: 20 * 1024 * 1024,
    });
    return await readFile(outPath);
  } catch (err) {
    if (err instanceof ExportError) throw err;
    const msg = err instanceof Error ? err.message : "pandoc export failed";
    throw new ExportError("PANDOC_FAILED", msg);
  } finally {
    await rm(dir, { recursive: true, force: true });
  }
}

export function exportContentType(format: ExportFormat): string {
  return FORMAT_META[format].mime;
}

File: crates/mp-rag/src/export.rs
// Copyright (c) 2026 Lane Ambrose
// SPDX-License-Identifier: MIT

//! Pandoc export for doctrine (docx/pdf). Fail-soft if pandoc is missing.

use std::process::Command;
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ExportError {
    InvalidFormat,
    PandocUnavailable,
    Empty,
    PandocFailed,
}

impl ExportError {
    pub fn code(&self) -> &'static str {
        match self {
            Self::InvalidFormat => "INVALID_FORMAT",
            Self::PandocUnavailable => "PANDOC_UNAVAILABLE",
            Self::Empty => "EMPTY",
            Self::PandocFailed => "PANDOC_FAILED",
        }
    }

    pub fn status(&self) -> u16 {
        match self {
            Self::InvalidFormat | Self::Empty => 400,
            Self::PandocUnavailable => 503,
            Self::PandocFailed => 502,
        }
    }
}

impl std::fmt::Display for ExportError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(match self {
            Self::InvalidFormat => "format must be docx or pdf",
            Self::PandocUnavailable => "pandoc is not installed or not on PATH",
            Self::Empty => "document is empty",
            Self::PandocFailed => "pandoc export failed",
        })
    }
}

pub fn is_pandoc_available() -> bool {
    Command::new("pandoc")
        .arg("--version")
        .stdout(std::process::Stdio::null())
        .stderr(std::process::Stdio::null())
        .status()
        .map(|s| s.success())
        .unwrap_or(false)
}

pub fn prepare_markdown_for_export(markdown: &str) -> String {
    let marker = "\n\n📎 Sources:";
    let cut = markdown.find(marker).unwrap_or(markdown.len());
    markdown[..cut].trim().to_string()
}

pub fn export_filename(source: &str, format: &str) -> String {
    let posix = source.replace('\\', "/");
    let file = posix.rsplit('/').next().unwrap_or("document");
    let base = file
        .trim_end_matches(".markdown")
        .trim_end_matches(".md");
    let ext = if format == "pdf" { "pdf" } else { "docx" };
    format!("{base}.{ext}")
}

pub fn export_content_type(format: &str) -> &'static str {
    if format == "pdf" {
        "application/pdf"
    } else {
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
    }
}

pub fn export_markdown(markdown: &str, format: &str) -> Result<Vec<u8>, ExportError> {
    if format != "docx" && format != "pdf" {
        return Err(ExportError::InvalidFormat);
    }
    if !is_pandoc_available() {
        return Err(ExportError::PandocUnavailable);
    }
    let body = prepare_markdown_for_export(markdown);
    if body.is_empty() {
        return Err(ExportError::Empty);
    }
    let n = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_nanos())
        .unwrap_or(0);
    let dir = std::env::temp_dir().join(format!("moneypenny-export-{n}"));
    let _ = std::fs::create_dir_all(&dir);
    let input = dir.join("input.md");
    let ext = if format == "pdf" { "pdf" } else { "docx" };
    let output = dir.join(format!("output.{ext}"));
    let write = (|| {
        std::fs::write(&input, body.as_bytes()).map_err(|_| ExportError::PandocFailed)?;
        let st = Command::new("pandoc")
            .arg(&input)
            .arg("-o")
            .arg(&output)
            .status()
            .map_err(|_| ExportError::PandocFailed)?;
        if !st.success() {
            return Err(ExportError::PandocFailed);
        }
        std::fs::read(&output).map_err(|_| ExportError::PandocFailed)
    })();
    let _ = std::fs::remove_dir_all(&dir);
    write
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn strips_sources_footer() {
        let s = prepare_markdown_for_export("Hello\n\n📎 Sources:\n- a");
        assert_eq!(s, "Hello");
    }

    #[test]
    fn filename_from_source() {
        assert_eq!(export_filename("ops/combat.md", "docx"), "combat.docx");
    }
}
